mycpp/gc_builtins.cc

```diff
@@ -335,114 +335,7 @@ Str* Str::replace(Str* old, Str* new_str) {
-    return NewList<Str*>(std::initializer_list<Str*>{kEmptyString});
-  }
-
-  // Find breaks first so we can allocate the right number of strings ALL AT
-  // ONCE. We want to avoid invalidating self->data_.
-  int num_bytes = 0;
-  int prev_pos = -1;
-  std::vector<int> breaks;
-  breaks.push_back(-1);  // beginning of first part
-
-  for (int i = 0; i < length; ++i) {
-    if (data_[i] == sep_char) {
-      breaks.push_back(i);
-      int part_len = i - prev_pos - 1;
-      if (part_len > 0) {  // only non-empty parts
-        num_bytes += aligned(kStrHeaderSize + part_len + 1);
-      }
-      prev_pos = i;
-    }
-  }
-  breaks.push_back(length);  // end of last part
-
-  if (length) {
-    int last_part_len = length - prev_pos - 1;
-    if (last_part_len > 0) {
-      num_bytes += aligned(kStrHeaderSize + last_part_len + 1);
-    }
-  }
-
-  result = NewList<Str*>(nullptr, breaks.size() - 1);  // reserve enough space
-
-  place = reinterpret_cast<char*>(gHeap.Allocate(num_bytes));
-  int n = breaks.size();
-  for (int i = 1; i < n; ++i) {
-    int prev_pos = breaks[i - 1];
-    int part_len = breaks[i] - prev_pos - 1;
-    if (part_len > 0) {
-      // like AllocStr(), but IN PLACE
-      int obj_len = kStrHeaderSize + part_len + 1;  // NUL terminator
-      Str* part = new (place) Str();                // placement new
-      part->SetObjLen(obj_len);                     // So the GC can copy it
-      memcpy(part->data_, self->data_ + prev_pos + 1, part_len);
-      result->set(i - 1, part);
-      place += aligned(obj_len);
-    } else {
-      result->set(i - 1, kEmptyString);  // save some space
-    }
-  }
-
-  return result;
-}
-
-Str* Str::join(List<Str*>* items) {
-  auto self = this;       // must be a root!
-  Str* result = nullptr;  // may not need to be a root, but make it robust
-
-  StackRoots _roots({&self, &result, &items});
-
-  int result_len = 0;
-  int num_parts = len(items);
-  if (num_parts == 0) {  // " ".join([]) == ""
-    return kEmptyString;
-  }
-  for (int i = 0; i < num_parts; ++i) {
-    result_len += len(items->index_(i));
-  }
-  int sep_len = len(self);
-  // add length of all the separators
-  result_len += sep_len * (num_parts - 1);
-
-  // log("len: %d", len);
-  // log("v.size(): %d", v.size());
-
-  result = AllocStr(result_len);
-  char* p_result = result->data_;  // advances through
-
-  for (int i = 0; i < num_parts; ++i) {
-    // log("i %d", i);
-    if (i != 0 && sep_len) {  // optimize common case of ''.join()
-      memcpy(p_result, self->data_, sep_len);  // copy the separator
-      p_result += sep_len;
-      // log("len_ %d", len_);
-    }
-
-    int n = len(items->index_(i));
-    if (n < 0) {
-      log("n: %d", n);
-      InvalidCodePath();
-    }
-    memcpy(p_result, items->index_(i)->data_, n);  // copy the list item
-    p_result += n;
-  }
-
-  assert(p_result[result_len] == '\0');  // GC should zero it
-  return result;
-}
+#include "mycpp/list_str_impl.h"
```

Comment:
everything in list_str_impl.h can be in `leaky_builtins.cc`

comment on Zulip: "leaky" really means "common between OLDSTL and GC"